src/main/java/com/guciowons/shoppingbasket/Product/Product.java

```diff
@@ -1,27 +1,41 @@
 package com.guciowons.shoppingbasket.Product;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+
 import java.math.BigDecimal;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.Objects;
 
+@Document
 public class Product {
 
-    private Integer id;
+    @Id
+    private String id;
+    private Integer externalId;
     private String title;
     private BigDecimal price;
     private String description;
-    private String category;
     private String image;
-    private Rating rating;
-    private Map<String, Object> additionalProperties = new HashMap<String, Object>();
 
-    public Integer getId() {
+    @JsonProperty("id")
```

Comment:
Dobra. W tej klasie mieszasz dwa bardzo różne obiekty, które mają różne odpowiedzialności, chociaż mogą nazywać się tak samo.

Jeden to Twój obiekt Product, który trzymasz w bazie i którym posługujesz się w swojej logice biznesowej.
Drugi to obiekt Product z zewnętrznego providera, którego używasz tylko do mapowania wyniku zapytania o produkty.

To powinny być dwie klasy! Co jeśli zewnętrzny dostawca zmieni pole "id" na "idChuj"? Wtedy będziesz musiał robić zmiany w swojej głównej logice, przez co uzależniasz cały swój program od tego dostawcy. Co jeśli będziesz chciał dodać wybór dla innych dostawców poduktów i mapping pól nie będzie się zgadzał?

Warstwa odpytywania o produkty powinna mieć swoje obiekty i przy pobieraniu ich powinna tłumaczyć je na Twój główny obiekt Product.

Mieszanie adnotacji mongo z tymi od mappera jsona nie jest dobrym pomysłem : )

Reply:
Najlepiej to by było jakbyś nie używał bezpośrednio `ProductClient` w `ProductService` tylko dopisał jeszcze jedną klasę pośredniczącą np `ProductProvider`, która właśnie weźmie produkty z zewnętrznego API i przetłumaczy je na twoje obiekty Product.